### ISSUE TITLE
Data race found in arch_init()

### ISSUE THREAD

**Issue by bbalp** [relic-toolkit/relic#228]
Hello,

Using `-DMULTI=PTHREAD` and TSAN I found a data race on this line:

https://github.com/relic-toolkit/relic/blob/7ef3af9e673da1fecf4677b8e73fadbfdb20f075/src/arch/relic_arch_x64.c#L61

The context is simple, when calling `core_init();` from multiple threads, the `core_ctx` variable for example is protected with `__thread`, but `arch_init();` is called too and `lzcnt_ptr` is not protected it's just declared `static`.

https://github.com/relic-toolkit/relic/blob/7ef3af9e673da1fecf4677b8e73fadbfdb20f075/src/arch/relic_arch_x64.c#L47

I think that it affects all the archs, and that it should be protected the same way `core_ctx` is, according the value set with `MULTI`.

Or maybe I am missing a critical information... What is your opinion?

**Comment by dfaranha** [comment]
Hi, thanks for the notification!

I understand in principle, but `lzcnt_ptr` is architecture-dependent and constant, so all threads should get the same value nonetheless?

**Comment by bbalp** [comment]
As I am not very knowledgeable on relic itself, when I got this TSAN report, I saw it wasn't symmetric with what's done regarding `core_ctx`, and reported the issue here. Because I prefer to have an expert opinion from someone like you to judge if there is something else I am missing, etc.

Now with your answer, I analysed a bit more the situation.

First I think we can agree that, `lzcnt_ptr` is let uninitialized when the program starts, until it is initialized in `arch_init()`, I haven't found anywhere else such initialization. So when you say it's constant you mean it's because it is guaranteed to always be the same value, whatever the thread writing into it.

From the point of view of the compiler. It is allowed to do whatever it wants, as long as it respects the constraints we put on our variables, etc. In our situation, the variable is meant to be written from multiple thread, with the same value on a global variable not protected of anything. Right now, I can't see how a compiler would be implemented such as it could cause a problem here ... but I am not writing compilers, and one lesson I learnt, is that when I am saying they are allowed to do anything they want, they are.

This paper at "2.4 Redundant writes" address this question in a general manner. https://www.usenix.org/legacy/events/hotpar11/tech/final%5ffiles/Boehm.pdf

Another argument would be, people are going to use TSAN more and more, and they expect the tools they are using to report no problems. TSAN is doing its job here, there is in fact multiple writes on the same unprotected variables.

Another argument would be, to maintain a safe code, let's keep it explicit regarding the intents of each operations so that if one day this variable is changing, it is explicit that there is a multi-threading case to take into account.

To sum it up, I think you are right there is little chance that this data race is harmful in any way. The only way to be absolutely certain is to know how compilers are implemented regarding this case. But yes I think it is way safer to stick to what the specifications are telling us to do in this situation.

Excellent answer from which I found the PDF I linked [https://stackoverflow.com/questions/23043251/parallel-writes-of-a-same-value](https://stackoverflow.com/questions/23043251/parallel-writes-of-a-same-value)

**Comment by dfaranha** [comment]
This makes a lot of sense, thank you! I guess the cleaner way to fix it is to make the pointer part of the context.

**Comment by dfaranha** [comment]
Which I just did in commit 2225e60189b7d06e74f6596f61132685efcbd582.

Hope it cleans the data race in TSAN!

**Comment by bbalp** [comment]
I will try and let you know. Thank you!

**Comment by bbalp** [comment]
I have tested your patch, and it's working as expected. Thank you for your work!
